tests/tsl/tsl3/test.cs

```diff
@@ -0,0 +1,268 @@
+                throw new ArgumentException();
+            if (request.Nums.Count == 10)
+                throw new InvalidDataException();
+            throw new NotImplementedException();
+        }
+
+        public override void TestAsynHandler(ReqReader request)
+        {
+            TestSynHandler(request);
+        }
+    }
+
+
+    public class TrinityServerFixture : IDisposable
+    {
+        public TrinityServerFixture()
+        {
+            Server = new TestServer();
+            Server.Start();
+        }
+
+        public TestServer Server { get; private set; }
+
+        public void Dispose()
+        {
+            Server.Stop();
+        }
+    }
+
+    [CollectionDefinition("TestServer Collection")]
+    public class TestServerCollection : ICollectionFixture<TrinityServerFixture>
+    {
+    }
+    #endregion
+
+    public static unsafe class Utils
+    {
+        public static byte* MakeCopyOfDataInReqWriter(ReqWriter writer)
+        {
+            byte* newbuf = (byte*)Memory.malloc((ulong)writer.Length);
+            Memory.memcpy(newbuf, writer.CellPtr, (ulong)writer.Length);
+            return newbuf;
+        }
+    }
+
+    public class StructTest
+    {
+        [Fact]
+        public void ReqStruct_TypeIsValueType()
+        {
+            Assert.True(typeof(Req).IsValueType);
+        }
+
+        [Fact]
+        public void ReqStruct_ShouldHaveDefaultCtor()
+        {
+            var req = new Req();
+        }
+
+        [Fact]
+        public void ReqStruct_ShouldHaveDefinedFields()
+        {
+            var req = new Req();
+            req.FieldBeforeList = 1984;
+            req.FieldAfterList = 42;
+            Assert.Null(req.Nums);
+            req.Nums = Enumerable.Concat(Enumerable.Repeat(2, 1), Enumerable.Repeat(3, 10)).ToList();
+        }
+
+        [Fact]
+        public void ReqStruct_FieldTypeShouldMatch()
+        {
+            var type = typeof(Req);
+            var fields = type.GetMembers().Where(m => m.MemberType == MemberTypes.Field).Cast<FieldInfo>().ToList();
+            Assert.True(fields.First(m => m.Name == "FieldBeforeList").FieldType == typeof(int), "FieldBeforeList");
+            Assert.True(fields.First(m => m.Name == "Nums").FieldType == typeof(List<int>), "Nums");
+            Assert.True(fields.First(m => m.Name == "FieldAfterList").FieldType == typeof(byte), "FieldAfterList");
+        }
+    }
+
+    public unsafe class MessageAccessorTest
+    {
+        [Fact]
+        public void Writer_ShouldBasicallyWork()
+        {
+            using (var writer = new ReqWriter())
+            {
+                writer.FieldBeforeList = -42;
+                writer.Nums.Add(100);
+                writer.FieldAfterList = 42;
+                Assert.Equal(writer.FieldBeforeList, -42);
+                Assert.Equal(writer.Nums.Count, 1);
+                Assert.Equal(writer.Nums[0], 100);
+                Assert.Equal(writer.FieldAfterList, 42);
```

Comment:
should also cast the writer into a struct and check the content, also, pls add some inline data rather than hard code it in.